CNV consensus (6 of 6): Merge consensus files and name columns

fingerfen opened the pull request:
<!--Hi there, thanks for your contribution! Please take a moment to fill out this template to facilitate the review of your pull request.-->

### Purpose/implementation Section

To add the last steps to the consensus calling pipeline. These steps merge any overlapping consensus segments and name the columns



#### What GitHub issue does your pull request address?

#128 

<!-- Check all those that apply or remove this section if it is not applicable.-->

- [ ] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [ ] This analysis has been added to continuous integration.
- [ ] This analysis is recorded in the table in `analyses/README.md`.

<!-- IF YOUR PULL REQUEST IS A DATA RELEASE, PLEASE REMOVE THE [HTML COMMENT TAG](https://html.com/tags/comment-tag/) FROM THE SECTION BELOW AND COMPLETE THE CHECKLIST-->

<!-- 
### Data Release Checklist
- [ ] Is the table in doc/data-file-descriptions.md up to date?
- [ ] Is doc/data-format.md up to date?
- [ ] Is doc/release-notes.md up to date?
- [ ] Is download-data.sh up to date?
-->

jashapiro commented on analyses/copy_number_consensus_call/Snakefile:
```diff
@@ -1,8 +1,6 @@
-               sample=config["samples"], 
-               combined_caller=["manta_cnvkit", "manta_freec", "cnvkit_freec"],
-               dupdel=["dup", "del"]),
+OUTPUT= expand("../../scratch/endpoints/{sample}.del_dup.merged.final.bed",
```
If these are the final output files, we should probably put them in a results directory within the analysis directory.

fingerfen replied:
You are right, I forgot to put them in the "result" folder. I have made the changes to reflect this.

fingerfen commented on analyses/copy_number_consensus_call/src/scripts/compare_variant_calling_updated.py:
```diff
@@ -233,8 +233,7 @@ def save_to_file(output_file_content, output_path, sample_name):
             ## Join the sample_name and single_name(file name) to the CNV info
-            k.extend([sample_name, single_name,'\n'])
-            file.write('\t'.join(k))
+            file.write('\t'.join(k) + '\t'+ sample_name + '\t' + single_name +'\n')
```
I applied your suggested change in my recent PR (part 5 of 6) and it output a `.bed` file fine. However, bedtools merge couldn't operate on that output file. It prompted that there were `"extra TAB at the end of the lines"`. When I inspected the file, there is an extra tab/indentation/character at the end of each line. When I do "cat -t {filename}", I see 
`chr3^I178470930^I184970385^I178470930..........sample^IBS_80078QDG.manta_cnvkit.del.bed^I`

That last `^I` is at the end of each line in the file and it prevents bedtools from operating.

With the change I made here, that `^I` is removed at the end of each line and thus allows bedtools to function.

jashapiro replied:
Yes, the previous version included the `\n` in the list, so it got joined with an extra `\t`. The version I just suggested does the join with only the fields, then adds the `\n` only in the write statement. It should not add an extra tab.

fingerfen replied:
I see, I missed that. I will apply this change then. Thanks for the clarification :)

fingerfen commented on analyses/copy_number_consensus_call/Snakefile:
```diff
@@ -161,4 +159,50 @@ rule compare_cnv_methods:
+    threads: 1
+    shell:
+        ## Merge the consensus files together
+        ## Put the dup CNVs calls and del CNVs calls together into one file
```
I definitely can. I will briefly explain it here to see if it makes sense for you. 

Columns 4, 5, and 6 hold the original CNV calls from Manta, CNVkit, and Freec, respectively. When we merge `manta_cnvkit, manta_freec, and cnvkit_freec` files together, `manta_cnvkit` and `manta_freec` will have information in column 4 (manta column). Thus, when using bedtools merging, we want to COLLAPSE these information and keep all information that is in column 4. The same applies to columns 5 and 6. 

Columns 7 and 8 are the CNVtype (DEL, DUP) and Sample_name, respectively. Among the `manta_cnvkit, manta_freec, and cnvkit_freec` files, **at this point**, we should have the **same** CNVtypes and Sample_name. Thus we perform DISTINCT, which is to take the unique of columns 7 and 8. 

As for column 9, this column holds the files that were merged to get a specific CNV. Thus we want to keep all information here so we COLLAPSE it.

jashapiro replied:
For columns 4, 5 and 6, will any of this result in duplication? For example, if the same original manta call was used in manta_cnvkit and manta_freec, will it now show up more than once? If so, I think we may want to clean that up in the final file, but that may require a bit of extra logic, so we would put it off to a separate PR.

fingerfen commented on analyses/copy_number_consensus_call/Snakefile:
```diff
@@ -161,4 +159,50 @@ rule compare_cnv_methods:
+    threads: 1
+    shell:
+        ## Name the headers
+        "python3 {input.script} --file {input.combined_dup_del} > {output.final_dup_del}"
```
There is no reason. It completely escaped my mind that I could do this using echo and cat. Your suggestion is way better. I will apply this change. Thanks again!

fingerfen commented:
@jashapiro, thank you for your review and suggestions. Once again, they were very helpful. I think I have applied all of the changes you suggested. Please have a look. I am currently working on the README.md and downloading the newest data release to run the entire pipeline on it from the beginning as suggested. I will let you know when the code finish running the newly downloaded data, hopefully by tomorrow. 

I will also be making changes to the methods.md file in a different PR as you suggested. 

Thank you

jashapiro commented:
Thanks @fingerfen, this all looks good from a first look, and it passed CI. Let me know if you run into any trouble with running it on the full data set, and I look forward to seeing the README when you get it done.

fingerfen commented:
@jashapiro The code finished running on my local system and 940 files were generated for DEL and 940 for DUP without any errors. Of course not all of these files have contents in them. These were then combined into one final file in `results/cnv_consensus.tsv`.

I also added the README.md. Please have a look.

Thanks!

fingerfen commented:
I am adding more to the README.md regarding the final output file

fingerfen commented on analyses/copy_number_consensus_call/README.md:
```diff
@@ -0,0 +1,43 @@
+This pipeline revolves around the use of Snakemake to run analysis for each patient sample. The overview of the steps are as followed:
+
+1) Parse through the 3 input files and put CNVs of the **same caller and sample** in the same files.
+2) Get rid of any one sample with **more than 2500** CNVs called
```
This only removes that one specific file for that caller as long as that file has more than 2500 CNVs called. As mentioned by @jharenza, we believe that files with more than 2500 CNVs called to be noisy/poor quality samples (this came from what GISTIC uses as a cutoff for noisy samples).

I will be adding this to the README.md

fingerfen commented:
I have updated the `analyses/README.md`. I have also made the suggested changes this analysis's README.md. I also added the final output file `cnv_consensus.tsv`. Please have a look. 

For some reason CI is failing, I checked the error and I believe it doesn't have anything to do with this analysis. I think I will have to wait for the error to be resolved.

jashapiro commented:
> For some reason CI is failing, I checked the error and I believe it doesn't have anything to do with this analysis. I think I will have to wait for the error to be resolved.
I think CI was just getting cancelled as you added each new change. The latest commit passed CI, so everything looks good on the CI front.

jashapiro commented:
I think we are really close here, and if CI passes, we can merge this commit. However, before we call this analysis "done" I would like to add one more step to your analysis to clean things up just a bit. 

I looked at the output file, and it does look like there is some duplicated information in there that could cause confusion down the line, but should be pretty straightforward to fix at this stage. 
In particular, when we look at the caller-specific CNVs used for each consensus call, there can be duplicated entries, as well as extraneous `NULL` values that we should probably delete.

For example, look at the consensus CNV
```
chr15	64681606	101072404	64635340:101704930:NA,64635340:101704930:NA,NULL,NULL,NULL	 ... BS_00TRPEQX
```
The manta_cnv calls there are two of the same location, followed by `NULL`. In the other callers there are also duplicated CNVs. 

It would be nice to write a script that parses through and removes duplicates and `NULL`s from fields 4,5 and 6 of each record to apply before combining all of the samples together into the final file.

If you have time to do that soon, let me know, or I can tackle it. To be clear, this would be a separate pull request (7 of 6, so to speak), as I think this one has enough in it already.

fingerfen commented:
@jashapiro  I agree that the NULL should be removed and I can write a script to address that no problem. My main concern here are the duplications. Thanks for catching that. It is not supposed to contain duplicated segments. I am curious as to why it has duplicated information and I would like to go back and find out the reason for it. I will take a look at it tonight and hopefully update you about the situation tomorrow.

fingerfen commented:
So I found out the reason why there are duplicates. It happens during the final bedtools merging step where columns 4, 5, and 6 get COLLAPSED together. In this step, there are 3 files that get merged, `manta-cnvkit`, `manta-freec`, and `cnvkit-manta`. I went back to these files to find any CNVs that overlapped with `chr15	   64681606	101072404`

When inspecting these files, the `manta-cnvkit` file contained a CNV that overlap with the coordinate above and it has `64635340:101704930:NA` in column 4. The `manta-freec` file also has 1 CNV with `64635340:101704930:NA`. The `cnvkit-manta` file contained 3 CNVs that overlaps with the coordinate above and they all have `NULL` in column 4. Since I merged these 3 files and COLLAPSED column 4, it resulted in `64635340:101704930:NA,64635340:101704930:NA,NULL,NULL,NULL`. This all makes sense.

I think a "solution" to this is to change COLLAPSE to DISTINCT in the merging step. However, I don't think this would get rid of the duplicated segments in more complicated situations. With this, I think it might be useful to make a script to remove duplicates and NULLs from columns 4, 5, and 6. 

Please let me know what you think! :)

jashapiro commented:
@fingerfen I suspected that was the cause, so I am glad to see you came to the same conclusion. I also agree that using `DISTINCT` would likely cause a different set of problems, so I think the correct solution is to include deduplication of those columns in a separate script along with NULL removal, after the bedtools merge.
I am going to approve this PR, so we can put that step in separately.
